cabal-install-3.6 fails to install network-3.1.2.2 in windows

* but cabal-install-3.4.0.0 is able to build it
* logs:

```
Configuring network-3.1.2.2...
configure: WARNING: unrecognized options: --with-compiler
checking build system type... x86_64-pc-msys
checking host system type... x86_64-pc-msys
checking for gcc... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe accepts -g... yes
checking for D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe option to accept ISO C89... none needed
checking for an ANSI C-conforming const... yes
checking how to run the C preprocessor... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe -E
checking for grep that handles long lines and -e... configure: error: no acceptable grep could be found in /d/ghcup/bin:/c/Users/atrey/AppData/Roaming/cabal/bin:/usr/bin:/mingw64/bin:/d/cabal/store/ghc-8.10.7/hsc2hs-0.68.7-640ce3a674b3ce528257d431f5e15fc40c5ab1a7/bin:/c/Program Files (x86)/Common Files/Oracle/Java/javapath:/c/WINDOWS/system32:/c/WINDOWS:/c/WINDOWS/System32/Wbem:/c/WINDOWS/System32/WindowsPowerShell/v1.0:/c/WINDOWS/System32/OpenSSH:/d/dev/app/git/cmd:/c/Users/atrey/AppData/Local/Microsoft/WindowsApps:/d/bin:/d/dev/ws/haskell/stack:/d/dev/app/emacs-26.2-x86_64/bin:/d/dev/app/Git/bin:/d/dev/app/curl_7_53_1_openssl_nghttp2_x64/bin:/d/dev/app/diffutils-2.8.7-1-bin/bin:/d/dev/app/grep-2.5.4-bin/bin:/d/dev/app/chocolatey/bin:/d/dev/app/vscode/bin:/d/Program Files/Java/jdk1.8.0_231/bin:/d/Program Files/7-Zip:/d/dev/app/node-v12.14.1-win-x64:/d/dev/ws/node.js:/d/dev/app/apache-maven-3.3.9/bin:/d/Program Files/PostgreSQL/11/bin:/d/bin/hls:/d/dev/app/python39/Scripts:/d/dev/app/python39:/d/ghcup/bin:/d/bin/cabal;/usr/xpg4/bin
cabal-3.6.0.0.exe: Failed to build network-3.1.2.2 (which is required by
exe:cabal from cabal-install-3.7.0.0). The failure occurred during the
configure step.
```

with cabal-3.4.0.0

```
Configuring network-3.1.2.2...
configure: WARNING: unrecognized options: --with-compiler
checking build system type... x86_64-pc-msys
checking host system type... x86_64-pc-msys
checking for gcc... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe accepts -g... yes
checking for D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe option to accept ISO C89... none needed
checking for an ANSI C-conforming const... yes
checking how to run the C preprocessor... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
....
```

Could it be related with #7510 (which i tested myself 🤦)?

Ok, i am afraid there had been some confusion that I could have avoided posting the full logs with versions of tooling:

```
PS D:\dev\ws\haskell\issues> cabal --version
cabal-install version 3.6.0.0
compiled using version 3.6.1.0 of the Cabal library
PS D:\dev\ws\haskell\issues> $(gcm cabal).path
D:\ghcup\bin\cabal.exe
PS D:\dev\ws\haskell\issues> cabal --store-dir D:\cabal\store2 install network --lib
Resolving dependencies...
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - hsc2hs-0.68.7 (exe:hsc2hs) (requires build)
 - network-3.1.2.2 (lib:network) (requires build)
Starting     hsc2hs-0.68.7 (exe:hsc2hs)
Building     hsc2hs-0.68.7 (exe:hsc2hs)
Installing   hsc2hs-0.68.7 (exe:hsc2hs)
Completed    hsc2hs-0.68.7 (exe:hsc2hs)
Starting     network-3.1.2.2 (all, legacy fallback)

Failed to build network-3.1.2.2. The failure occurred during the configure
step.
Build log (
D:\cabal\logs\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39.log
):
Configuring network-3.1.2.2...
configure: WARNING: unrecognized options: --with-compiler
checking build system type... x86_64-pc-msys
checking host system type... x86_64-pc-msys
checking for gcc... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe accepts -g... yes
checking for D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe option to accept ISO C89... none needed
checking for an ANSI C-conforming const... yes
checking how to run the C preprocessor... D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe -E
checking for grep that handles long lines and -e... configure: error: no acceptable grep could be found in /d/ghcup/bin:/c/Users/atrey/AppData/Roaming/cabal/bin:/usr/bin:/mingw64/bin:/d/cabal/store2/ghc-8.10.7/hsc2hs-0.68.7-640ce3a674b3ce528257d431f5e15fc40c5ab1a7/bin:/c/Program Files (x86)/Common Files/Oracle/Java/javapath:/c/WINDOWS/system32:/c/WINDOWS:/c/WINDOWS/System32/Wbem:/c/WINDOWS/System32/WindowsPowerShell/v1.0:/c/WINDOWS/System32/OpenSSH:/d/dev/app/git/cmd:/c/Users/atrey/AppData/Local/Microsoft/WindowsApps:/d/bin:/d/dev/ws/haskell/stack:/d/dev/app/emacs-26.2-x86_64/bin:/d/dev/app/Git/bin:/d/dev/app/curl_7_53_1_openssl_nghttp2_x64/bin:/d/dev/app/diffutils-2.8.7-1-bin/bin:/d/dev/app/grep-2.5.4-bin/bin:/d/dev/app/chocolatey/bin:/d/dev/app/vscode/bin:/d/Program Files/Java/jdk1.8.0_231/bin:/d/Program Files/7-Zip:/d/dev/app/node-v12.14.1-win-x64:/d/dev/ws/node.js:/d/dev/app/apache-maven-3.3.9/bin:/d/Program Files/PostgreSQL/11/bin:/d/bin/hls:/d/dev/app/python39/Scripts:/d/dev/app/python39:/d/ghcup/bin:/d/bin/cabal;/usr/xpg4/bin
cabal-3.6.0.0.exe: Failed to build network-3.1.2.2. See the build log above
for details.
```

with `-v3` we can see the configure invocation:

```
"D:\ghcup\msys64\usr\bin\sh.exe" "C:\/TEMP/cabal-install.-6000/dist-newstyle/tmp/src-6000/network-3.1.2.2/configure" "--with-compiler=ghc" "--prefix=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39" "--bindir=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39\bin" "--libdir=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39\lib" "--libexecdir=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39\libexec" "--datadir=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39\share" "--sysconfdir=D:\cabal\store2\ghc-8.10.7\network-3.1.2.2-43da37077e8a7b3eeb1dbd1af4e0673f5a10eb39\etc" "CC=D:\ghcup\ghc\8.10.7\lib\../mingw/bin\gcc.exe"
```

> [Edit2: after reading #7494, I agree there should be no colons, but only semicolons in PATH. Co cabal is giving wrong PATH to configure or the following not always holds: "When cabal is called the msys2-runtime converts the environment block to contain windows paths since Cabal is a Windows program and expect a Windows style environmental block."]

It does, but the problem is that the `PATH` environment variable is special. At startup `sh` will convert whatever `PATH` it inherits.. since the shell itself uses `:`.

So Cabal has Windows Path -> sh (converts everything) 

It looks like we have two seemingly incompatible requirements. 2.70 requires different handling for `ACLOCAL_PATH` but `PATH` looks except from translations..

The problem seems to be that the packaged `network` comes with an `configure` script generated from the old version of `autoconf` while (at least me) tested with `network` github which requires you to run `autoreconf` to generate a new `configure`.

So the two are seemingly incompatible.

It looks like @phadej was right, we need a more complicated fix..

There are likely 3 options:
1. lobby for an upstream fix (unlikely)
2. lobby for msys2 to revert https://git.savannah.gnu.org/cgit/autoconf.git/commit/?id=35a1c64600894ecc3b06b6c4b273952db7c8fc83 down stream.
3. Make cabal send unix paths instead of Windows paths to all PATHs it sends to configure (including the path to configure scripts themselves).

But the last time I tried 3 the error became even more cryptic 
```
configure: error: cannot find required auxiliary files 
```
with an empty files list.  Can be worth trying again though.  I am also worried that paths exported end up in configure output. 

So we can try just modifying the path to configure scripts to unix paths and see if that works if not I think 2 is likely the only option left.

> configure: error: cannot find required auxiliary files 

I have debugged the same issue a few months ago in the GHC build. The relevant ticket/comment is https://gitlab.haskell.org/ghc/ghc/-/issues/19189#note_332168

It led me to remove the use of `config.sub/config.guess` in `base` and `ghc-bignum` (https://gitlab.haskell.org/ghc/ghc/-/commit/bcaa36c4277dc577863132da2f95645715e2a15e). Sadly this fix can't be easily generalized to other packages...

I have opened this issue upstream at the time (https://savannah.gnu.org/support/?110448) which got the following responses by email:

>>> But $PATH_SEPARATOR=":"
>>
>> Intuitively, I would expect PATH_SEPARATOR to be ";" on Windows (unless you're using CygWin).
>
>Indeed.  And if you are using Cygwin, shouldn't you be using /cygdrive/c/... instead of c:\... ?  (or however it's spelled, it's been a very long time since I used Cygwin)

--- a/Cabal/src/Distribution/Simple.hs
+++ b/Cabal/src/Distribution/Simple.hs
@@ -1,6 +1,7 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE CPP #-}
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  Distribution.Simple
@@ -97,7 +98,11 @@ import Distribution.System (buildPlatform)
 import System.Environment (getArgs, getProgName)
 import System.Directory   (removeFile, doesFileExist
                           ,doesDirectoryExist, removeDirectoryRecursive)
-import System.FilePath                      (searchPathSeparator, takeDirectory, (</>), splitDirectories, dropDrive)
+import System.FilePath    (searchPathSeparator, takeDirectory, (</>),
+                           splitDirectories, dropDrive)
+#ifdef mingw32_HOST_OS
+import System.FilePath    (normalise, splitDrive)
+#endif
 import Distribution.Compat.ResponseFile (expandResponse)
 import Distribution.Compat.Directory        (makeAbsolute)
 import Distribution.Compat.Environment      (getEnvironment)
@@ -670,7 +675,7 @@ runConfigureScript verbosity backwardsCompatHack flags lbi = do
   -- TODO: We don't check for colons, tildes or leading dashes. We
   -- also should check the builddir's path, destdir, and all other
   -- paths as well.
-  let configureFile' = intercalate "/" $ splitDirectories configureFile
+  let configureFile' = toUnix configureFile
   for_ badAutoconfCharacters $ \(c, cname) ->
     when (c `elem` dropDrive configureFile') $
       warn verbosity $ concat
@@ -709,6 +714,19 @@ runConfigureScript verbosity backwardsCompatHack flags lbi = do
                ++ "Unix compatibility toolchain such as MinGW+MSYS or Cygwin. "
                ++ "If you are not on Windows, ensure that an 'sh' command "
                ++ "is discoverable in your path."
+
+-- | Convert Windows path to Unix ones
+toUnix :: String -> String
+#ifdef mingw32_HOST_OS
+toUnix s = let tmp = normalise s
+               (l, rest) = case splitDrive tmp of
+                             ([],  x) -> ("/"      , x)
+                             (h:_, x) -> ('/':h:"/", x)
+               parts = splitDirectories rest
+           in  l ++ intercalate "/" parts
+#else
+toUnix s = intercalate "/" $ splitDirectories s
+#endif
 
 badAutoconfCharacters :: [(Char, String)]
 badAutoconfCharacters =

